text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,224 @@
+and thus is ready for migrating to CDK `V2`,
+you can set the `CDK_DEPRECATED_ERROR` environment variable,
+which will make any CDK command that invokes synthesis
+(`cdk synth`, `cdk deploy`, `cdk diff`, etc.)
+fail with an exception if any deprecated element is used.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+We are doing this to help customers have a smoother migration from `V1` to `V2` of the CDK.
+
+### Why should we _not_ do this?
+
+I see three reasons why we might not want to implement this feature:
+
+1. Deprecated elements are already highlighted by virtually any editor/IDE.
+  In my experience, customers are diligent in moving away from deprecated APIs,
+  so this feature might not provide much value.
+2. We risk overloading customers with warnings,
+  training them to ignore our output if there's too much of it.
+  A warning must be read in order to be effective, after all.
+3. Depending on the exact solution chosen
+  (see discussion below),
+  we might not be able to distinguish from the user's code using a deprecated API,
+  and from our own libraries using deprecated APIs.
+  Which means a user might get warnings that they will be unable to get rid of
+  (because they're coming from code that they don't control).
+
+### What changes are required to enable this change?
+
+There are two high-level components of this change:
+
+#### 1. Changes to JSII
+
+We will add a feature to JSII that adds the warning code to the emitted JavaScript for deprecated elements.
+It will be off by default, and will have to be activated explicitly to take effect.
+We will also add two options to the JSII CLI that allow passing the names of the environment variables used for,
+respectively, silencing the warnings, and turning the warnings into errors.
+
+#### 2. Using the changed JSII
+
+Once we have modified JSII and released a new version of it,
+we will need to use it in the CDK.
+We will have to start compiling CDK with the new option turned on,
+and also set the two additional options mentioned above to be
+`CDK_DEPRECATED_IGNORE` and `CDK_DEPRECATED_ERROR`, respectively.
+
+We should also modify our CDK test infrastructure to run with the `CDK_DEPRECATED_ERROR` option turned on by default,
+unless a test is explicitly checking a deprecated API as a regression test,
+in which case we should add an API for turning that option off for those tests.
+
+### Is this a breaking change?
+
+No.
+
+### What are the drawbacks of this solution?
+
+1. Requires changes to JSII.
+2. Does not allow distinguishing between the customer using deprecated APIs,
+  and the library code using deprecated APIs.
+  We can alleviate this problem by turning on the `CDK_DEPRECATED_ERROR` environment variable by default in our tests,
+  which will be a forcing function for us to stop using deprecated APIs in the Construct Library.
+3. We won't be able to register warnings for accessing deprecated static constants
```

Comment:
Not actually true I don't think.

We can turn:

```ts
class Foo {
  /** @deprecated */
  public static readonly CONSTANT = 5;
}
// into
class Foo {
  public static get CONSTANT(): number {
    console.error('[WARNING] oh no');
    return 5;
  }
}
```

And something similar for enums.

Honestly, I feel like something like:

```ts
// Would work fine and would work for enums too
Foo.prototype.CONSTANT = deprecatedAccessor(Foo.prototype.CONSTANT, 'Use Foo.Bloop instead');
```

Reply:
Fair enough. That's a huge intrusion to the generated code though. I've re-phrased the downside to reflect that.